### PR TITLE
Show Loader when adding Workspace Edit

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details
- Add loading flags for Avatar upload and form submission
- IMP: Added `isUploading` flag to AvatarWithImagePicker with uploadIndicator
- Replaced js loop to Animated.loop

### Fixed Issues

$ https://github.com/Expensify/App/issues/4487

### Tests
1. Tested by keeping the form contents the same
2. Updating the workspace name and image
3. Removing the workspace image
4. Submitting the form successfully
5. Uploading a large image

### QA Steps
1. Go to Workspace and click on the Avatar
2. Check button state when the values are not updated
3. Change the content/upload the image and see if values are updated
3. Test Removing the workspace image as well 
4. Submitting the form successfully
5. Uploading a large image

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
**v1**
https://user-images.githubusercontent.com/3069065/131385161-e70fc702-9ef7-4515-bacd-86ac728d5c03.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132609914-524caa8b-ead7-4ec1-97c4-6288d6575dd7.mp4

#### Mobile Web
**v1**
https://user-images.githubusercontent.com/3069065/131384341-a5d4da09-e5e7-4807-8cf9-3724c797d2c6.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610024-500c89e5-83d0-4c05-91c6-61688fd223c1.mp4

#### Desktop
**v1**
https://user-images.githubusercontent.com/3069065/131384466-c488f178-844d-471b-b9b6-9ca68fa5de65.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610083-e81b6468-c976-4ec2-83cc-f7932de947f3.mp4

#### iOS
**v1**
https://user-images.githubusercontent.com/3069065/131384510-c38701d8-74ae-49c2-a116-e50e2fdc19a5.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610283-1a851e62-640f-496d-86d2-83376a6e7125.mp4

#### Android
**v1**
https://user-images.githubusercontent.com/3069065/131384679-c6b796b0-f49e-4e76-8d48-3991d2cac241.mp4

**Updated**
https://user-images.githubusercontent.com/3069065/132610438-29e4b86c-81aa-45fc-ad41-7d77e0ba6c92.mp4

### CODE REVIEW

**Comment by mananjadhav** [comment]
@roryabraham I've raised a draft PR. Need to clarify a couple of things:

1. Should we throw error messages on any form error or avatar upload? or even a success message?
2. I've noticed when we upload an Avatar image, there isn't any feedback shown to the user. For now, what I've done is added an `isUploading` flag to `AvatarWithImagePicker` which shows the `Sync` icon instead of Pencil when set true. While it did solve the purpose, it doesn't look great. I would recommend an animation similar to `AvatartWithIndicator`. Shall I go ahead and add it?
3. If we agree to do the loader thing in #2, a similar UX is missing for the Profile image update as well. Shall I cover that too in this PR if is a 🟢 for the above approach.

**Comment by roryabraham** [comment]
> Should we throw error messages on any form error or avatar upload? or even a success message

If there's a form error or API error, we should display a user-friendly error message. A success message would be good too, maybe using [Growl](https://github.com/Expensify/App/blob/2bde5c1ce2c6563676e47ff061f4dd4861bf1a19/src/libs/Growl.js) or some other visual indication.

> I've noticed when we upload an Avatar image, there isn't any feedback shown to the user. For now, what I've done is added an isUploading flag to AvatarWithImagePicker which shows the Sync icon instead of Pencil when set true. While it did solve the purpose, it doesn't look great. I would recommend an animation similar to AvatartWithIndicator. Shall I go ahead and add it?

I'm not sure. I'll defer to @Expensify/design here

> If we agree to do the loader thing in Fix spaces #2, a similar UX is missing for the Profile image update as well. Shall I cover that too in this PR if is a 🟢 for the above approach?

I assume you meant item 2 of your list, not the PR linked in the comment. I would think yes 🟢, but again will defer to @Expensify/design here.

---

Lastly, I'll warn you that what I see in this PR so far violates the [Expensify/App Philosophy](https://github.com/Expensify/App#philosophy), specifically `UI Binds to Disk`. My interpretation of that philosophy is that actions should never really be returning a promise – instead the UI binds to data in Onyx, and calls actions that manage the data. It should never be explicitly waiting for a response.

**Comment by mananjadhav** [comment]
Thanks for the clarification @roryabraham.


 
> If there's a form error or API error, we should display a user-friendly error message. A success message would be good too, maybe using [Growl](https://github.com/Expensify/App/blob/2bde5c1ce2c6563676e47ff061f4dd4861bf1a19/src/libs/Growl.js) or some other visual indication.

I'll put in a message.

> I'm not sure. I'll defer to @Expensify/design here

 I'll wait for the comments from the design team.

> I assume you meant item 2 of your list, not the PR linked in the comment. 

Yes. I meant point# 2 from my list. It seems GH linked to the PR due to #.

> Lastly, I'll warn you that what I see in this PR so far violates the [Expensify/App Philosophy](https://github.com/Expensify/App#philosophy), specifically `UI Binds to Disk`. 

Thanks for highlighting.  I only went ahead with the approach because the same file had the following line:

```
this.uploadAvatarPromise = uploadAvatar(image).then(url => new Promise((resolve) => {
```
So should I store these loading props to Onyx?

**Comment by roryabraham** [comment]
> So should I store these loading props to Onyx?

For now, I believe this is the approach we'll want to take.

**Comment by mananjadhav** [comment]
@roryabraham I've updated the PR with the following changes:
1. Growl messages in catch block
2. Moving all flags to Onyx
3. Uploading Indicator in the AvatarImagePicker


I haven't added Growl on success because immediately after the policy is saved, we redirect to workspace card with updated info. Let me know if you still feel we should add a Growl message for success.

**Comment by roryabraham** [comment]
@mananjadhav No growl on success sounds fine to me 👍

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (stitesExpensify) [inline src/pages/workspace/WorkspaceEditorPage.js:59]: We can prevent repeating this line by putting in in an `always()` right?

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (stitesExpensify) [inline src/pages/workspace/WorkspaceEditorPage.js:82]: Since there isn't one of these in the `then()` how does the `isPolicyUpdating` get changed when the promise does not fail?

**Review by stitesExpensify — CHANGES_REQUESTED** [review]

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (mananjadhav) [reply src/pages/workspace/WorkspaceEditorPage.js:59]: Can you elaborate this please?

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (mananjadhav) [reply src/pages/workspace/WorkspaceEditorPage.js:82]: It's in the `then` block of `update` function in the API call.

Comment (stitesExpensify) [reply src/pages/workspace/WorkspaceEditorPage.js:82]: Gotcha 👍

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (mananjadhav) [reply src/pages/workspace/WorkspaceEditorPage.js:59]: Got it. Sure. Will update.

**Comment by mananjadhav** [comment]
@stitesExpensify @roryabraham I've updated the PR. Additionally, I've also modified the size of the indicator to get it closed to the Avatar and adjusted the icon size.


https://user-images.githubusercontent.com/3069065/131907746-2cb1af94-8dd4-45c3-a338-ddc49f27863e.mov

File: src/pages/workspace/WorkspaceEditorPage.js
Comment (mananjadhav) [reply src/pages/workspace/WorkspaceEditorPage.js]: Done

**Review by stitesExpensify — DISMISSED** [review]
LGTM! All you @roryabraham

File: src/components/AvatarWithImagePicker.js
Comment (roryabraham) [inline src/components/AvatarWithImagePicker.js:108]: Rather than using recursion here, it's better to use `Animated.loop` because it won't block the JS thread to make the recursive call:

```javascript
Animated.loop(
    Animated.timing(this.rotate, {
        toValue: 1,
        duration: 2000,
        easing: Easing.linear,
        isInteraction: false,
        useNativeDriver: true,
    }),
).start();
```

File: src/components/AvatarWithImagePicker.js
Comment (roryabraham) [inline src/components/AvatarWithImagePicker.js:219]: Spacing is just a little weird here (should be 4 spaces past the `<>` not 2):

```suggestion
                                  {
```

File: src/components/AvatarWithImagePicker.js
Comment (roryabraham) [inline src/components/AvatarWithImagePicker.js:220]: Also, this should probably be indented 4 spaces past the enclosing `{}`:

Like this:

```javascript
{
    this.props.isUploading
        ? ()
        : ()
}
```

**Review by roryabraham — CHANGES_REQUESTED** [review]
Just had a few small comments, overall this looks great!

File: src/components/AvatarWithImagePicker.js
Comment (roryabraham) [reply src/components/AvatarWithImagePicker.js:108]: i.e: the animation will restart immediately in the UI thread, instead of:

1. Sending a message through the bridge back to JS to say the animation is done
1. JS thread is blocked to execute the callback
1. Callback sends a message back through the bridge to UI thread to restart the animation.

Comment (mananjadhav) [reply src/components/AvatarWithImagePicker.js:108]: Thanks for the clarification. I haven't worked with Animation earlier. It seems though we're using the same thing in `AvatarWithIndicator`. What should we do about that?

File: src/components/AvatarWithImagePicker.js
Comment (mananjadhav) [reply src/components/AvatarWithImagePicker.js:219]: Will check. Not sure how editor didn't pick this.

File: src/components/AvatarWithImagePicker.js
Comment (mananjadhav) [reply src/components/AvatarWithImagePicker.js]: Updated and works great. Thanks for the suggestion. 🙌

File: src/components/AvatarWithImagePicker.js
Comment (mananjadhav) [reply src/components/AvatarWithImagePicker.js]: Fixed this

File: src/components/AvatarWithImagePicker.js
Comment (mananjadhav) [reply src/components/AvatarWithImagePicker.js:217]: @roryabraham It seems the `eslint` rolls back to two spaces when we use `<>`.  I also checked across the app, it seems that we're using two spaces only after `<>`. 

image
image
image

Please let me know if you meant anything else.

**Comment by mananjadhav** [comment]
PR updated

File: src/components/AvatarWithImagePicker.js
Comment (roryabraham) [reply src/components/AvatarWithImagePicker.js:217]: Okay, no worries about the spacing then, we can just trust ESLint.

**Review by roryabraham — DISMISSED** [review]
Thanks @mananjadhav, looks good. If you wanted to update AvatarWithIndicator to use Animated.loop for consistency and a small performance boost that would be great!

Someday I hope we can switch over to using Reanimated instead of React Native's Animated. It comes with much better animation performance and in my opinion a simpler API. But that's a conversation for another day 😄

**Comment by mananjadhav** [comment]
I’ll update the AvatarWithIndicator too. 

Yeah I’ve worked on Reanimated v1. But I’ve heard Reanimated V2 last year reduces a lot of boilerplate code too.

**Comment by mananjadhav** [comment]
PR updated

**Review by roryabraham — CHANGES_REQUESTED** [review]
Okay, I apologize for not noticing these things earlier, but I have a few more comments:

### Start both animations in parallel
I just noticed that there are two animations that are basically happening in parallel, so we should just use `Animated.parallel`, something like this:

```javascript
/**
 * Start Animation for Indicator
 *
 * @memberof AvatarWithImagePicker
 */
start() {
    Animated.loop(
        Animated.parallel([
            Animated.timing(this.rotate, {
                toValue: 1,
                duration: 2000,
                easing: Easing.linear,
                isInteraction: false,
                useNativeDriver: true,
            }),
            Animated.spring(this.scale, {
                toValue: 1.666,
                tension: 1,
                isInteraction: false,
                useNativeDriver: true,
            }),
        ]),
    ).start();
}

/**
 * Stop Animation for Indicator
 *
 * @memberof AvatarWithImagePicker
 */
stop() {
    Animated.spring(this.scale, {
        toValue: 1,
        tension: 1,
        isInteraction: false,
        useNativeDriver: true,
    }).start(() => {
        this.rotate.resetAnimation();
        this.scale.resetAnimation();
        this.rotate.setValue(0);
    });
}
```

### DRY up animation code

To do this, I think we should:

1. Create a new file at `src/styles/animations/SpinningIndicatorAnimation.js`
1. That file should export a single class called `SpinningIndicatorAnimation`:

    ```javascript
      import {Animated, Easing} from 'react-native';

	  class SpinningIndicatorAnimation {
	      constructor() {
	          this.rotate = new Animated.Value(0);
	          this.scale = new Animated.Value(1);
	          this.start = this.start.bind(this);
	          this.stop = this.stop.bind(this);
	          this.getSyncingStyles = this.getSyncingStyles.bind(this);
	      }
	  
	      /**
	       * Start Animation for Indicator
	       *
	       * @memberof AvatarWithImagePicker
	       */
	      start() {
	          Animated.loop(
	              Animated.parallel([
	                  Animated.timing(this.rotate, {
	                      toValue: 1,
	                      duration: 2000,
	                      easing: Easing.linear,
	                      isInteraction: false,
	                      useNativeDriver: true,
	                  }),
	                  Animated.spring(this.scale, {
	                      toValue: 1.666,
	                      tension: 1,
	                      isInteraction: false,
	                      useNativeDriver: true,
	                  }),
	              ]),
	          ).start();
	      }
	  
	      /**
	       * Stop Animation for Indicator
	       *
	       * @memberof AvatarWithImagePicker
	       */
	      stop() {
	          Animated.spring(this.scale, {
	              toValue: 1,
	              tension: 1,
	              isInteraction: false,
	              useNativeDriver: true,
	          }).start(() => {
	              this.rotate.resetAnimation();
	              this.scale.resetAnimation();
	              this.rotate.setValue(0);
	          });
	      }
	  
	      /**
	       * Get Indicator Styles while animating
	       *
	       * @returns {Object}
	       */
	      getSyncingStyles() {
	          return {
	              transform: [{
	                  rotate: this.rotate.interpolate({
	                      inputRange: [0, 1],
	                      outputRange: ['0deg', '-360deg'],
	                  }),
	              },
	              {
	                  scale: this.scale,
	              }],
	          };
	      }
	  }
	  
	  export default SpinningIndicatorAnimation;
    ```
    
1. Then both `AvatarWithImagePicker` and `AvatarWithIndicator` can just do:

    ```javascript
    constructor(props) {
        super(props);
        this.animation = new SpinningIndicatorAnimation();
        ...
        ...
    }
    ```
    
    And then replace all uses of the animation methods with just `this.animation.start()`, `this.animation.stop()`, or `this.animation.getSyncingStyles`.

### Scaling issue
Lastly, I'm looking at the screenshots provided in the PR description, and it doesn't seem like the indicator is getting bigger when the animation starts. Instead, it actually looks like it's getting smaller and is outside the avatar circle.

**Comment by mananjadhav** [comment]
Okay, I'll check the Animation updates and implement them.

> Lastly, I'm looking at the screenshots provided in the PR description, and it doesn't seem like the indicator is getting bigger > when the animation starts. Instead, it actually looks like it's getting smaller and is outside the avatar circle.

I had updated this comment but not the PR description. Let me update it in the PR description.

https://github.com/Expensify/App/pull/4801#issuecomment-912008496

**Comment by mananjadhav** [comment]
@roryabraham There's a comment in `AvatarWithIndicator.startRotation`,

```
We need to manually loop the animations as `useNativeDriver` does not work well with `Animated.loop`.
```

**Comment by mananjadhav** [comment]
@roryabraham I am updating the `SpinnerAnimation` and having the following issue with `AvatarWithIndicator`, which I can't reproduce in staging.new.expensify.com

If I add a couple of messages in Offline mode, and then disable throttling, the `Avatar` scale doesn't change back to 1. (stays 1.66 and hence it looks bigger), neither does the rotation stop. I've pushed my change and attaching the video of the issue.


https://user-images.githubusercontent.com/3069065/132413898-d7d3abb1-168b-4c01-af82-75e828a6f056.mov


I am guessing it is because after `this.animation.stop()` the component isn't rendered and `getIndicatorStyles` don't have the updated values of `this.scale` and `this.rotate`.  I also feel `Animated.parallel` use must be causing it. I am a little confused that I am unable to reproduce the same in the staging URL where we have the old Animation in place.

**Comment by mananjadhav** [comment]
@roryabraham Did you get a chance to look at my comments above?

Meanwhile I'll resolve the conflicts.

**Comment by roryabraham** [comment]
@mananjadhav Regarding this comment:

> We need to manually loop the animations as `useNativeDriver` does not work well with `Animated.loop`.

is there some observable problem that occurs when you use `Animated.loop` as I suggested above? I looked back at the PR which added that comment and there wasn't any additional context explaining the problem. AFAICT there's nothing that seems to go wrong when I switch to using `Animated.loop` locally, and it makes the code clearer and more performant. So I'd prefer to do that unless we can somehow expand on how "`useNativeDriver` does not work well with `Animated.loop`". Hopefully that makes sense!

We can skip the use of `Animated.parallel` though since you've found an issue with my suggested implementation.

**Comment by mananjadhav** [comment]
Well, I didn't find any issue on the emulator but I haven't been able to check on a real Android device. Let me see if I can test and accordingly update the PR.

**Comment by mananjadhav** [comment]
@roryabraham PR updated. I checked on a real Android device and it worked fine. So I am guessing it shouldn't be a problem.

I've also updated the videos with the latest testing of indicator closer to the image with increased size.

**Review by roryabraham — APPROVED** [review]
Okay, thanks for making all those changes, this is looking great!

**Comment by botify** [comment]
This has been deployed to production and is now subject to a 7-day regression period.
If no regressions arise, payment will be issued on 2021-09-17.  :confetti_ball:

**Pull request merged**
